homeassistant/components/snapcast/media_player.py

```diff
@@ -61,51 +77,111 @@ async def async_setup_entry(
+            known_ids -= ids_to_remove
+            known_ids |= ids_to_add
 
+            return ids_to_add, ids_to_remove
 
-async def handle_set_latency(entity, service_call):
-    """Handle the entity service set_latency."""
-    if not isinstance(entity, SnapcastClientDevice):
-        raise TypeError("Latency can only be set for a Snapcast client.")
-    await entity.async_set_latency(service_call.data[ATTR_LATENCY])
+        new_entities: list[SnapcastBaseDevice] = []
 
+        group_ids = {g.identifier for g in coordinator.server.groups}
+        groups_to_add, groups_to_remove = _update_known_ids(_known_group_ids, group_ids)
 
-class SnapcastGroupDevice(MediaPlayerEntity):
-    """Representation of a Snapcast group device."""
+        new_entities.extend(
+            [
+                SnapcastGroupDevice(
+                    coordinator, coordinator.server.group(group_id), host_id
+                )
+                for group_id in groups_to_add
+            ]
+        )
+
+        client_ids = {c.identifier for c in coordinator.server.clients}
+        clients_to_add, clients_to_remove = _update_known_ids(
+            _known_client_ids, client_ids
+        )
+
+        new_entities.extend(
+            [
+                SnapcastClientDevice(
+                    coordinator, coordinator.server.client(client_id), host_id
+                )
+                for client_id in clients_to_add
+            ]
+        )
+
+        _LOGGER.debug(
+            "New clients: %s",
+            str([coordinator.server.client(c).friendly_name for c in clients_to_add]),
+        )
+        _LOGGER.debug(
+            "New groups: %s",
+            str([coordinator.server.group(g).friendly_name for g in groups_to_add]),
+        )
+        _LOGGER.debug(
+            "Remove client IDs: %s",
+            str([list(clients_to_remove)]),
+        )
+        _LOGGER.debug(
+            "Remove group IDs: %s",
+            str(list(groups_to_remove)),
+        )
+
+        # Add new entities
+        async_add_entities(new_entities)
+
+        # Remove stale entities
+        entity_registry = er.async_get(hass)
+        for group_id in groups_to_remove:
+            if entity_id := entity_registry.async_get_entity_id(
+                MEDIA_PLAYER_DOMAIN,
+                DOMAIN,
+                SnapcastGroupDevice.get_unique_id(host_id, group_id),
+            ):
+                entity_registry.async_remove(entity_id)
+
+        for client_id in clients_to_remove:
+            if entity_id := entity_registry.async_get_entity_id(
+                MEDIA_PLAYER_DOMAIN,
+                DOMAIN,
+                SnapcastClientDevice.get_unique_id(host_id, client_id),
+            ):
+                entity_registry.async_remove(entity_id)
+
+    coordinator.async_add_listener(_check_entities)
+
+    # Remove any existing entities
+    entity_registry = er.async_get(hass)
+    entity_registry.async_clear_config_entry(config_entry.entry_id)
```

Comment:
euh why?

Reply:
While testing I found its possible to get the Snapserver and HA out of sync if a group/client disappears while HA is offline. (e.g. a group is removed while HA is restarting).

HA still has entities for these groups, but since their IDs aren't in `_known_group_ids` they are not removed on update.

Clearing the entities was simplest solution. Otherwise you'd have to rebuild the list of known IDs.

Reply:
You are right, this is an issue with the current implementation, too. Could you read known ids from the registry instead?

Reply:
The client/group IDs could be parsed from the unique IDs if necessary. Is there any harm in starting from a blank state?

Reply:
I'd rather not rely on that. What you can do is check all the devices you have after you forwarded the setups to the platforms, and compare them with all the devices you have for that config entry in the device registry

Reply:
> I'd rather not rely on that. What you can do is check all the devices you have after you forwarded the setups to the platforms

Sorry, I'm not following. At this point aren't I already in the platform setup?


> compare them with all the devices you have for that config entry in the device registry

I still have the problem that the entities in the device registry don't have the Snapcast group/client ID I need (unless parsed from the unique ID)

Reply:
No I mean, this code can live in `__init__.py` and there you have a call where you forward the setups to the platforms. After that is completed, all devices are present. So if you then take all the devices you have in the device registry for this config entry, you can compare it with the data you have, and remove the ones you don't have data for.

But this is bigger than just a few lines, so I would recommend doing this in a separate PR

Reply:
Sorry, I might need this spelled out.

> No I mean, this code can live in __init__.py and there you have a call where you forward the setups to the platforms.

There was an earlier review comment that specifically wanted this code moved into the platform (i.e. `media_player.py`).

> So if you then take all the devices you have in the device registry for this config entry, you can compare it with the data you have, and remove the ones you don't have data for.

This would still require parsing the unique IDs to determine the group/client ID. I took your previous comments to indicate this would be undesirable.

Reply:
> There was an earlier review comment that specifically wanted this code moved into the platform (i.e. media_player.py).

Could it be that this was about the creation?

> I took your previous comments to indicate this would be undesirable.

If you could link that I can maybe I can elaborate (or learn from that it can be recieved differently)

But I think if we remove it we can merge this and leave the device removal for a different PR discussion

Reply:
> Could it be that this was about the creation?

I suppose. :) Based on the examples it looked like the entire entity lifecycle was handled in the platform.



> If you could link that I can maybe I can elaborate (or learn from that it can be recieved differently)

This was the comment that I interpreted to mean "don't parse the unique IDs".
![image](https://github.com/user-attachments/assets/602c2ac2-6d36-4d51-aaab-07e7fd657b70)

https://github.com/home-assistant/core/pull/124499#discussion_r1820761762


> But I think if we remove it we can merge this and leave the device removal for a different PR discussion

Understood.

Thanks for your patience.

Reply:
Gotcha.

I've made the requested change for now.